<h2>Análise de dados com pandas - Olist</h2>

In [1]:
#Importação pandas
import pandas as pd
#Importação Datetime
import datetime as dt

Importação de arquivos para DataFrames.

In [2]:
df_order_items_dataset = pd.read_csv('data\order_items_dataset.csv')
df_orders_dataset = pd.read_csv('data\orders_dataset.csv')
df_customers_dataset = pd.read_csv('data\customers_dataset.csv')
df_order_payments_dataset = pd.read_csv('data\order_payments_dataset.csv')
df_sellers_dataset = pd.read_csv('data\sellers_dataset.csv')

<h4>Acessando DataFrames para melhor compreensão do dados fornecidos. </4>

Pedidos de compra

In [3]:
df_orders_dataset.head()

order_id                       customer_id  \
0  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
1  53cdb2fc8bc7dce0b6741e2150273451  b0830fb4747a6c6d20dea0b8c802d7ef   
2  47770eb9100c2d0c44946d9cf07ec65d  41ce2a54c0b03bf3443c3d931a367089   
3  949d5b44dbf5de918fe9c16f97b45f8a  f88197465ea7920adcdbec7375364d82   
4  ad21c59c0840e6cb83a9ceb5573f8159  8ab97904e6daea8866dbdbc4fb7aad2c   

  order_status order_purchase_timestamp    order_approved_at  \
0    delivered      2017-10-02 10:56:33  2017-10-02 11:07:15   
1    delivered      2018-07-24 20:41:37  2018-07-26 03:24:27   
2    delivered      2018-08-08 08:38:49  2018-08-08 08:55:23   
3    delivered      2017-11-18 19:28:06  2017-11-18 19:45:59   
4    delivered      2018-02-13 21:18:39  2018-02-13 22:20:29   

  order_delivered_carrier_date order_delivered_customer_date  \
0          2017-10-04 19:55:00           2017-10-10 21:25:13   
1          2018-07-26 14:31:00           2018-08-07 15:27:45   
2          2018-08-08 13:50:00           2018-08-17 18:06:29   
3          2017-11-22 13:39:59           2017-12-02 00:28:42   
4          2018-02-14 19:46:34           2018-02-16 18:17:02   

  order_estimated_delivery_date  
0           2017-10-18 00:00:00  
1           2018-08-13 00:00:00  
2           2018-09-04 00:00:00  
3           2017-12-15 00:00:00  
4           2018-02-26 00:00:00

Inclui os itens adquiridos em cada pedido de compra 

In [4]:
df_order_items_dataset.head()

order_id  order_item_id  \
0  00010242fe8c5a6d1ba2dd792cb16214              1   
1  00018f77f2f0320c557190d7a144bdd3              1   
2  000229ec398224ef6ca0657da4fc703e              1   
3  00024acbcdf0a6daa1e931b038114c75              1   
4  00042b26cf59d7ce69dfabb4e55b4fd9              1   

                         product_id                         seller_id  \
0  4244733e06e7ecb4970a6e2683c13e61  48436dade18ac8b2bce089ec2a041202   
1  e5f2d52b802189ee658865ca93d83a8f  dd7ddc04e1b6c2c614352b383efe2d36   
2  c777355d18b72b67abbeef9df44fd0fd  5b51032eddd242adc84c38acab88f23d   
3  7634da152a4610f1595efa32f14722fc  9d7a1d34a5052409006425275ba1c2b4   
4  ac6c3623068f30de03045865e4e10089  df560393f3a51e74553ab94004ba5c87   

   shipping_limit_date   price  freight_value  
0  2017-09-19 09:45:35   58.90          13.29  
1  2017-05-03 11:05:13  239.90          19.93  
2  2018-01-18 14:48:30  199.00          17.87  
3  2018-08-15 10:10:18   12.99          12.79  
4  2017-02-13 13:57:51  199.90          18.14

Informação geográfica sobre os clientes

In [5]:
df_customers_dataset.head()

customer_id                customer_unique_id  \
0  06b8999e2fba1a1fbc88172c00ba8bc7  861eff4711a542e4b93843c6dd7febb0   
1  18955e83d337fd6b2def6b18a428ac77  290c77bc529b7ac935b93aa66c333dc3   
2  4e7b3e00288586ebd08712fdd0374a03  060e732b5b29e8181a18229c7b0b2b5e   
3  b2b6027bc5c5109e529d4dc6358b12c3  259dac757896d24d7702b9acbbff3f3c   
4  4f2d8ab171c80ec8364f7c12e35b23ad  345ecd01c38d18a9036ed96c73b8d066   

   customer_zip_code_prefix          customer_city customer_state  
0                     14409                 franca             SP  
1                      9790  sao bernardo do campo             SP  
2                      1151              sao paulo             SP  
3                      8775        mogi das cruzes             SP  
4                     13056               campinas             SP

Métodos de pagamentos, valores e parcelas

In [6]:
df_order_payments_dataset.head()

order_id  payment_sequential payment_type  \
0  b81ef226f3fe1789b1e8b2acac839d17                   1  credit_card   
1  a9810da82917af2d9aefd1278f1dcfa0                   1  credit_card   
2  25e8ea4e93396b6fa0d3dd708e76c1bd                   1  credit_card   
3  ba78997921bbcdc1373bb41e913ab953                   1  credit_card   
4  42fdf880ba16b47b59251dd489d4441a                   1  credit_card   

   payment_installments  payment_value  
0                     8          99.33  
1                     1          24.39  
2                     1          65.71  
3                     8         107.78  
4                     2         128.45

Informação geográfica sobre os vendedores

In [7]:
df_sellers_dataset.head()

seller_id  seller_zip_code_prefix  \
0  3442f8959a84dea7ee197c632cb2df15                   13023   
1  d1b65fc7debc3361ea86b5f14c68d2e2                   13844   
2  ce3ad9de960102d0677a81f5d0bb7b2d                   20031   
3  c0f3eea2e14555b6faeea3dd58c1b1c3                    4195   
4  51a04a8a6bdcb23deccc82b0b80742cf                   12914   

         seller_city seller_state  
0           campinas           SP  
1         mogi guacu           SP  
2     rio de janeiro           RJ  
3          sao paulo           SP  
4  braganca paulista           SP

<h4> Tratamento de dados</h4>

Checar se há linhas com valores nulos nos dataframes. Caso haja, então ocorre exclusão destes.

In [8]:
df_orders_dataset.isnull().values.any()

True

In [9]:
df_order_items_dataset.isnull().values.any()

False

In [10]:
df_customers_dataset.isnull().values.any()

False

In [11]:
df_order_payments_dataset.isnull().values.any()

False

In [12]:
df_sellers_dataset.isnull().values.any()

False

In [13]:
df_orders_dataset = df_orders_dataset.dropna()

<h4>Levantamento de premissas e hipóteses. </h4>

<h4>1. Análise da data de entrega e data estimada, se estimada for maior que a entregue, então está atrasada. Pegar ID dos pedidos atrasados para descobrir quem é o vendedor responsável pelo atraso. Checar quantas vezes o ID do vendedor aparece em entregas atrasadas. Checar quantidade de entregas atrasadas em relação à quantidade de vendas totais.  </h4>

Comparação entre a data de entrega e a data estimada pra a entrega. As colunas resultantes da comparação foram guardadas no novo DataFrame 'df_late_deliveries'.

In [14]:
df_late_deliveries = df_orders_dataset.query('order_delivered_customer_date > order_estimated_delivery_date')

Criação de outro Dataframe para uso futuro.

In [15]:
df_seasonality = df_late_deliveries.copy()

Junção do Dataframe 'df_late_deliveries' com a 'df_order_items_dataset' para obtenção do vendedor relacionado às entregas atrasadas encontradas. 

In [16]:
df_late_results = pd.merge(df_late_deliveries, df_order_items_dataset, how='left', on='order_id')

In [17]:
df_late_results = df_late_results.drop_duplicates(subset=['order_id'])

Exclusão de colunas desnecessárias para análise. 

In [18]:
df_late_results = df_late_results.drop('order_status', axis=1)
df_late_results = df_late_results.drop('order_purchase_timestamp', axis=1)
df_late_results = df_late_results.drop('order_approved_at', axis=1)
df_late_results = df_late_results.drop('order_delivered_carrier_date', axis=1)
df_late_results = df_late_results.drop('order_delivered_customer_date', axis=1)
df_late_results = df_late_results.drop('order_item_id', axis=1)
df_late_results = df_late_results.drop('order_estimated_delivery_date', axis=1)
df_late_results = df_late_results.drop('price', axis=1)
df_late_results = df_late_results.drop('shipping_limit_date', axis=1)
df_late_results = df_late_results.drop('freight_value', axis=1)

In [19]:
df_late_results.head()

order_id                       customer_id  \
0  203096f03d82e0dffbc41ebc2e2bcfb7  d2b091571da224a1b36412c18bc3bbfe   
1  fbf9ac61453ac646ce8ad9783d7d0af6  3a874b4d4c4b6543206ff5d89287f0c3   
2  8563039e855156e48fccee4d611a3196  5f16605299d698660e0606f7eae2d2f9   
3  6ea2f835b4556291ffdc53fa0b3b95e8  c7340080e394356141681bd4c9b8fe31   
4  66e4624ae69e7dc89bd50222b59f581f  684fa6da5134b9e4dab731e00011712d   

                         product_id                         seller_id  
0  5ac9d9e379c606e36a8094a6046f75dc  633ecdf879b94b5337cca303328e4a25  
1  7b717060aa783eb7f23a747a3a733dd7  c0563dd588b775f2e37747ef6ad6c92c  
2  bff2010b28e8fbcff5a9db9d3fea5ac4  955fee9216a65b617aa5c0531780ce60  
3  be021417a6acb56b9b50d3fd2714baa8  f5f46307a4d15880ca14fab4ad9dfc9b  
4  b37b72d5a56f887725c2862184b8cab8  db4350fd57ae30082dec7acbaacc17f9

Checagem de existência de ID`s de vendedores duplicadas.

In [20]:
boolean = df_late_results['seller_id'].duplicated().any()
print(boolean, end='\n\n')

True



Contagem de total de vendas por vendedor.

In [21]:
df_sellers_info =  df_order_items_dataset['seller_id'].value_counts()

Montagem de Dataframe com as informações.

In [22]:
df_sellers_info = df_sellers_info.to_frame()

Correção de index e nome de colunas.

In [23]:
df_sellers_info = df_sellers_info.reset_index()

In [24]:
df_sellers_info = df_sellers_info.rename(columns = {'index': 'seller_id', 'seller_id': 'total_sellings'})

In [25]:
df_sellers_info.head()

seller_id  total_sellings
0  6560211a19b47992c3666cc44a7e94c0            2033
1  4a3ca9315b744ce9f8e9374361493884            1987
2  1f50f920176fa81dab994f9023523100            1931
3  cc419e0650a3c5ba77189a1882b7556a            1775
4  da8622b14eb17ae2831f4ac5b9dab84a            1551

Contagem de vendas com entrega atrasada por vendedor.

In [26]:
df_sellers_info_right = df_late_results['seller_id'].value_counts()

Montagem de Dataframe com as informações.

In [27]:
df_sellers_info_right = df_sellers_info_right.to_frame()

Correção de index e nome de colunas.

In [28]:
df_sellers_info_right = df_sellers_info_right.reset_index()

In [29]:
df_sellers_info_right = df_sellers_info_right.rename(columns = {'index': 'seller_id', 'seller_id': 'total_late'})

In [30]:
df_sellers_info_right.head()

seller_id  total_late
0  4a3ca9315b744ce9f8e9374361493884         194
1  1f50f920176fa81dab994f9023523100         148
2  4869f7a5dfa277a7dca6462dcf3b52b2         130
3  ea8482cd71df3c1969d7b9473ff13abc         118
4  6560211a19b47992c3666cc44a7e94c0         116

Junção de Dataframes para obter a relação do total de vendas e total de entregas atrasadas de cada vendedor.

In [31]:
df_late_results_final = pd.merge(df_sellers_info, df_sellers_info_right, how='left', on='seller_id')

Exclusão de vendedores sem entregas atrasadas.

In [32]:
df_late_results_final = df_late_results_final.dropna()

In [33]:
df_late_results_final.head()

seller_id  total_sellings  total_late
0  6560211a19b47992c3666cc44a7e94c0            2033       116.0
1  4a3ca9315b744ce9f8e9374361493884            1987       194.0
2  1f50f920176fa81dab994f9023523100            1931       148.0
3  cc419e0650a3c5ba77189a1882b7556a            1775       101.0
4  da8622b14eb17ae2831f4ac5b9dab84a            1551       100.0

<span style="color:red">Proposta de solução:</span> Primeiramente, deve ser realizado um estudo do motivo dos atrasos, que pode ser desde ausência de planejamento, falta de produto em estoque, ou inúmeros outros fatores. Após a compreensão disso, devem ser tomada medidas que reparem estes atrasos recorrentes dos mesmos vendedores.

 <h4>2. O atraso está ligado a diferença de estado entre vendedor e cliente? </h4>

Captura do Dataframe de entregas atrasadas 'df_late_results' já utilizado anteriormente, pois contém as informações necessárias: ID`s do vendedor e do cliente.

In [34]:
df_late_results.head()

order_id                       customer_id  \
0  203096f03d82e0dffbc41ebc2e2bcfb7  d2b091571da224a1b36412c18bc3bbfe   
1  fbf9ac61453ac646ce8ad9783d7d0af6  3a874b4d4c4b6543206ff5d89287f0c3   
2  8563039e855156e48fccee4d611a3196  5f16605299d698660e0606f7eae2d2f9   
3  6ea2f835b4556291ffdc53fa0b3b95e8  c7340080e394356141681bd4c9b8fe31   
4  66e4624ae69e7dc89bd50222b59f581f  684fa6da5134b9e4dab731e00011712d   

                         product_id                         seller_id  
0  5ac9d9e379c606e36a8094a6046f75dc  633ecdf879b94b5337cca303328e4a25  
1  7b717060aa783eb7f23a747a3a733dd7  c0563dd588b775f2e37747ef6ad6c92c  
2  bff2010b28e8fbcff5a9db9d3fea5ac4  955fee9216a65b617aa5c0531780ce60  
3  be021417a6acb56b9b50d3fd2714baa8  f5f46307a4d15880ca14fab4ad9dfc9b  
4  b37b72d5a56f887725c2862184b8cab8  db4350fd57ae30082dec7acbaacc17f9

Criação de Dataframe que apresente o estado em que o vendedor envia os produtos.

In [35]:
df_seller_state = pd.merge(df_late_results, df_sellers_dataset, how='left', on='seller_id')

Verificação seguida de remoção de ID`s duplicados dos vendedores.

In [36]:
boolean = df_seller_state['seller_id'].duplicated().any()
print(boolean, end='\n\n')

True



In [37]:
df_seller_state = df_seller_state.drop_duplicates(subset=['seller_id'])

Criação de Dataframe que apresente os estados em que o vendedor envia os produtos e que os clientes os recebem, respectivamente.

In [38]:
df_customer_state = pd.merge(df_seller_state, df_customers_dataset, how='left', on='customer_id')

Remoção de colunas desnecessárias para a análise.

In [39]:
df_customer_state = df_customer_state.drop('seller_zip_code_prefix', axis=1)
df_customer_state = df_customer_state.drop('seller_city', axis=1)
df_customer_state = df_customer_state.drop('customer_unique_id', axis=1)
df_customer_state = df_customer_state.drop('customer_zip_code_prefix', axis=1)
df_customer_state = df_customer_state.drop('customer_city', axis=1)

In [40]:
df_customer_state.head()

order_id                       customer_id  \
0  203096f03d82e0dffbc41ebc2e2bcfb7  d2b091571da224a1b36412c18bc3bbfe   
1  fbf9ac61453ac646ce8ad9783d7d0af6  3a874b4d4c4b6543206ff5d89287f0c3   
2  8563039e855156e48fccee4d611a3196  5f16605299d698660e0606f7eae2d2f9   
3  6ea2f835b4556291ffdc53fa0b3b95e8  c7340080e394356141681bd4c9b8fe31   
4  66e4624ae69e7dc89bd50222b59f581f  684fa6da5134b9e4dab731e00011712d   

                         product_id                         seller_id  \
0  5ac9d9e379c606e36a8094a6046f75dc  633ecdf879b94b5337cca303328e4a25   
1  7b717060aa783eb7f23a747a3a733dd7  c0563dd588b775f2e37747ef6ad6c92c   
2  bff2010b28e8fbcff5a9db9d3fea5ac4  955fee9216a65b617aa5c0531780ce60   
3  be021417a6acb56b9b50d3fd2714baa8  f5f46307a4d15880ca14fab4ad9dfc9b   
4  b37b72d5a56f887725c2862184b8cab8  db4350fd57ae30082dec7acbaacc17f9   

  seller_state customer_state  
0           SP             SP  
1           SP             RJ  
2           SP             TO  
3           SC             SP  
4           SP             SE

Criação de Dataframe que contém a diferença entre os estados dos vendedores e dos clientes.

In [41]:
df_different_states = df_customer_state.query('seller_state != customer_state')

In [42]:
df_different_states.head()

order_id                       customer_id  \
1  fbf9ac61453ac646ce8ad9783d7d0af6  3a874b4d4c4b6543206ff5d89287f0c3   
2  8563039e855156e48fccee4d611a3196  5f16605299d698660e0606f7eae2d2f9   
3  6ea2f835b4556291ffdc53fa0b3b95e8  c7340080e394356141681bd4c9b8fe31   
4  66e4624ae69e7dc89bd50222b59f581f  684fa6da5134b9e4dab731e00011712d   
6  6a0a8bfbbe700284feb0845d95e0867f  68451b39b1314302c08c65a29f1140fc   

                         product_id                         seller_id  \
1  7b717060aa783eb7f23a747a3a733dd7  c0563dd588b775f2e37747ef6ad6c92c   
2  bff2010b28e8fbcff5a9db9d3fea5ac4  955fee9216a65b617aa5c0531780ce60   
3  be021417a6acb56b9b50d3fd2714baa8  f5f46307a4d15880ca14fab4ad9dfc9b   
4  b37b72d5a56f887725c2862184b8cab8  db4350fd57ae30082dec7acbaacc17f9   
6  f8a8f05a35976a91aed5cccc3992c357  4a3ca9315b744ce9f8e9374361493884   

  seller_state customer_state  
1           SP             RJ  
2           SP             TO  
3           SC             SP  
4           SP             SE  
6           SP             RJ

Total de atrasos em entregas gerais.

In [43]:
df_customer_state.shape

(1388, 6)

Total de atrasos em entregas interestaduais.

In [44]:
df_different_states.shape

(984, 6)

<span style="color:red">Proposta de solução:</span> Foi possível analisar que as entregas interestaduais compõem parcela majoritária das entregas atrasadas em geral. Isso corresponde a aproximadamente 70.9% das entregas em atraso (984). Sabendo disso, é importante que haja uma reavaliação da definição dos prazos. 

 <h4>3. Checar se há maior taxa de atraso com sazonalidade (natal e black friday) </h4>

Tratamento de coluna de datas e agrupamento de pedidos atrasados por mês e ano. 

In [45]:
df_seasonality['order_delivered_customer_date'] = pd.to_datetime(df_seasonality['order_delivered_customer_date'])

In [46]:
df_seasonality = df_seasonality['order_id'].groupby([df_seasonality['order_delivered_customer_date'].dt.year.rename('year'), df_seasonality['order_delivered_customer_date'].dt.month.rename('month')]).agg({'count'})

Correção de index.

In [47]:
df_seasonality = df_seasonality.reset_index()

Exposição de entregas atrasadas referentes ao ano de 2017.

In [48]:
df_seasonality.loc[df_seasonality['year'] == 2017]

year  month  count
2   2017      2      2
3   2017      3     64
4   2017      4    103
5   2017      5    226
6   2017      6    114
7   2017      7    117
8   2017      8    119
9   2017      9    175
10  2017     10    219
11  2017     11    242
12  2017     12    765

<span style="color:red">Proposta de solução:</span> Foi realizada a análise apenas do ano de 2017, pois era o ano mais completo do conjunto de dados. Enquanto 2017 apresenta 11 meses disponíveis, 2016 e 2018 apresentam, respectivamente, 2 e 10 meses. Por meio da análise foi possível perceber que a quantidade de atrasos do mês de Maio, Outubro, Novembro e Dezembro foram superiores em relação aos outros meses do ano. Isso, provavelmente, se dá pelas datas comemorativas (dia das mães, dia das crianças, black friday e natal). Portanto, deve ocorrer um melhor planejamento para que se atenda às demandas destas datas com maior pontualidade. 

Seria possível levantar inúmeras hipóteses e propostas, porém, as análises aqui neste case foram realizadas sob a ótica de um só viés, os atrasos de entregas, que foram considerados por mim os que mais necessitam de atenção.